In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

df = pd.read_csv('Data/FC26_20250921.csv')
print("Shape:", df.shape)
df.head()

Shape: (18405, 110)


C:\Users\ENG_ALI_WENINY\AppData\Local\Temp\ipykernel_10876\704209732.py:7: DtypeWarning: Columns (0: player_tags) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('Data/FC26_20250921.csv')


,player_id,player_url,fifa_version,fifa_update,fifa_update_date,short_name,long_name,player_positions,overall,potential,...,cdm,rdm,rwb,lb,lcb,cb,rcb,rb,gk,player_face_url
0,252371,/player/252371/jude-bellingham/260004/,26,4,2025-09-19,J. Bellingham,Jude Victor William Bellingham,"CAM, CM",90,94,...,85+3,85+3,83+3,82+3,81+3,81+3,81+3,82+3,18+3,https://cdn.sofifa.net/players/252/371/26_120.png
1,239053,/player/239053/federico-valverde/260004/,26,4,2025-09-19,F. Valverde,Federico Santiago Valverde Dipetta,"CM, CDM, RB",89,90,...,87+3,87+3,86+3,86+3,83+3,83+3,83+3,86+3,18+3,https://cdn.sofifa.net/players/239/053/26_120.png
2,212622,/player/212622/joshua-kimmich/260004/,26,4,2025-09-19,J. Kimmich,Joshua Walter Kimmich,"CDM, RB, CM",89,89,...,87+2,87+2,86+3,85+3,82+3,82+3,82+3,85+3,21+3,https://cdn.sofifa.net/players/212/622/26_120.png
3,235212,/player/235212/achraf-hakimi/260004/,26,4,2025-09-19,A. Hakimi,Achraf Hakimi Mouhأشرف حكيمي,"RB, RM",89,90,...,83+3,83+3,86+3,86+3,81+3,81+3,81+3,86+3,17+3,https://cdn.sofifa.net/players/235/212/26_120.png
4,224232,/player/224232/nicolo-barella/260004/,26,4,2025-09-19,N. Barella,Nicolò Barella,CM,87,87,...,85+2,85+2,84+3,83+3,80+3,80+3,80+3,83+3,19+3,https://cdn.sofifa.net/players/224/232/26_120.png


In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 18405 entries, 0 to 18404
Columns: 110 entries, player_id to player_face_url
dtypes: float64(16), int64(48), str(46)
memory usage: 15.4 MB


In [4]:
df.describe()

,player_id,fifa_version,fifa_update,overall,potential,value_eur,wage_eur,age,height_cm,weight_kg,...,mentality_composure,defending_marking_awareness,defending_standing_tackle,defending_sliding_tackle,goalkeeping_diving,goalkeeping_handling,goalkeeping_kicking,goalkeeping_positioning,goalkeeping_reflexes,goalkeeping_speed
count,18405.000000,18405.0,18405.0,18405.000000,18405.000000,1.840500e+04,18405.000000,18405.000000,18405.000000,18405.000000,...,18405.000000,18405.000000,18405.000000,18405.000000,18405.000000,18405.000000,18405.000000,18405.000000,18405.000000,2062.000000
mean,207378.375822,26.0,4.0,65.766965,71.165173,2.931633e+06,10096.055420,25.222548,182.001358,75.141755,...,57.747786,46.253572,48.435914,46.248737,16.285683,16.082749,15.998642,16.139147,16.343928,34.489816
std,73761.964821,0.0,0.0,6.980628,6.403862,7.947787e+06,20082.981891,4.773553,6.891484,6.891777,...,12.257447,20.723894,21.038765,20.575523,17.589875,17.002686,16.770157,17.162219,17.859712,10.552502
min,19541.000000,26.0,4.0,47.000000,49.000000,0.000000e+00,0.000000,16.000000,155.000000,47.000000,...,15.000000,5.000000,7.000000,6.000000,2.000000,2.000000,2.000000,2.000000,2.000000,15.000000
25%,195859.000000,26.0,4.0,61.000000,67.000000,4.750000e+05,1000.000000,21.000000,177.000000,70.000000,...,50.000000,28.000000,29.000000,26.000000,8.000000,8.000000,8.000000,8.000000,8.000000,26.000000
50%,237681.000000,26.0,4.0,66.000000,71.000000,1.000000e+06,4000.000000,25.000000,182.000000,75.000000,...,59.000000,52.000000,56.000000,53.000000,11.000000,11.000000,11.000000,11.000000,11.000000,33.000000
75%,260601.000000,26.0,4.0,70.000000,75.000000,2.100000e+06,10000.000000,29.000000,187.000000,80.000000,...,66.000000,63.000000,65.000000,63.000000,14.000000,14.000000,14.000000,14.000000,14.000000,43.000000
max,280142.000000,26.0,4.0,91.000000,95.000000,1.745000e+08,610000.000000,44.000000,210.000000,105.000000,...,93.000000,91.000000,91.000000,89.000000,90.000000,90.000000,91.000000,90.000000,90.000000,65.000000


In [5]:
df.isnull().sum().sort_values(ascending=False)

work_rate               18405
nation_position         17677
nation_team_id          17677
nation_jersey_number    17677
player_tags             17455
                        ...  
cb                          0
rcb                         0
rb                          0
gk                          0
player_face_url             0
Length: 110, dtype: int64

In [6]:
cols_to_drop = ['work_rate', 'nation_position', 'nation_team_id', 'nation_jersey_number',
                 'player_tags', 'goalkeeping_speed', 'player_traits', 'release_clause_eur']
df = df.drop(columns=cols_to_drop)

df['is_goalkeeper'] = df['player_positions'].str.contains('GK').astype(int)
df = df[df['is_goalkeeper'] == 0].copy()
df = df.drop(columns=['is_goalkeeper'])


gk_cols = ['goalkeeping_diving', 'goalkeeping_handling', 'goalkeeping_kicking',
           'goalkeeping_positioning', 'goalkeeping_reflexes']
df = df.drop(columns=gk_cols)

main_stats = ['pace', 'shooting', 'passing', 'dribbling', 'defending', 'physic']
df[main_stats] = df[main_stats].fillna(0)


df['club_joined_date'] = pd.to_datetime(df['club_joined_date'], errors='coerce')
df['years_at_club'] = (pd.Timestamp('2025-09-19') - df['club_joined_date']).dt.days / 365
df['years_at_club'] = df['years_at_club'].fillna(df['years_at_club'].median())
df = df.drop(columns=['club_joined_date'])

df['contract_years_remaining'] = df['club_contract_valid_until_year'] - 2025
df = df.drop(columns=['club_contract_valid_until_year'])

df = df.dropna(subset=['club_name'])

df['is_loaned'] = df['club_loaned_from'].notna().astype(int)
df = df.drop(columns=['club_loaned_from'])

extra_cols_to_drop = ['player_url', 'long_name', 'dob', 'real_face', 'player_face_url',
                       'body_type', 'club_position',
                       'ls','st','rs','lw','lf','cf','rf','rw','lam','cam','ram','lm',
                       'lcm','cm','rcm','rm','lwb','ldm','cdm','rdm','rwb','lb','lcb','cb','rcb','rb','gk']
df = df.drop(columns=extra_cols_to_drop)

id_cols_to_drop = ['player_id', 'fifa_version', 'fifa_update', 'fifa_update_date',
                    'short_name', 'league_id', 'club_team_id', 'nationality_id',
                    'club_jersey_number']
df = df.drop(columns=id_cols_to_drop)

df = df.drop(columns=['club_name'])


print(df.shape)
null_counts = df.isnull().sum()

(16266, 53)


In [7]:
df.dtypes

player_positions                   str
overall                          int64
potential                        int64
value_eur                        int64
wage_eur                         int64
age                              int64
height_cm                        int64
weight_kg                        int64
league_name                        str
league_level                   float64
nationality_name                   str
preferred_foot                     str
weak_foot                        int64
skill_moves                      int64
international_reputation         int64
pace                           float64
shooting                       float64
passing                        float64
dribbling                      float64
defending                      float64
physic                         float64
attacking_crossing               int64
attacking_finishing              int64
attacking_heading_accuracy       int64
attacking_short_passing          int64
attacking_volleys        

In [8]:
def group_position(pos):
    pos = pos.split(',')[0].strip()
    if pos in ['CB', 'LB', 'RB', 'LWB', 'RWB']:
        return 'DEF'
    elif pos in ['CDM', 'CM', 'CAM', 'LM', 'RM']:
        return 'MID'
    elif pos in ['ST', 'CF', 'LW', 'RW']:
        return 'ATT'
    else:
        return 'OTHER'

df['position_group'] = df['player_positions'].apply(group_position)
df = df.drop(columns=['player_positions'])
print(df['position_group'].value_counts())

position_group
MID    6832
DEF    6088
ATT    3346
Name: count, dtype: int64


In [9]:
print(df['league_name'].nunique())
print(df['nationality_name'].nunique())

42
158


In [10]:
nationality_freq = df['nationality_name'].value_counts()
df['nationality_freq'] = df['nationality_name'].map(nationality_freq)
df = df.drop(columns=['nationality_name'])

In [11]:
league_freq = df['league_name'].value_counts()
df['league_freq'] = df['league_name'].map(league_freq)
df = df.drop(columns=['league_name'])

In [12]:
df = pd.get_dummies(df, columns=['preferred_foot', 'position_group'], drop_first=True)

In [ ]:
print(df.shape)
print(df.dtypes.value_counts())

(16266, 54)
int64      42
float64     9
bool        3
Name: count, dtype: int64


In [ ]:
os.makedirs('Data', exist_ok=True)
df.to_csv('Data/clean_data.csv', index=False)


لحد هنا 110 بقوا 54 عمود 
والصفوف قلوا حوالي 2200 شيلنا الحراس والفري اجينت
